flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/ttl/RocksDbTtlCompactFilterUtils.java

```diff
@@ -0,0 +1,197 @@
+		if (enableTtlCompactionFilter && metaInfoBase instanceof RegisteredKeyValueStateBackendMetaInfo) {
+			RegisteredKeyValueStateBackendMetaInfo kvMetaInfoBase = (RegisteredKeyValueStateBackendMetaInfo) metaInfoBase;
+			if (TtlStateFactory.TtlSerializer.isTtlStateSerializer(kvMetaInfoBase.getStateSerializer())) {
+				return createAndSetCompactFilterFactory(options);
+			}
+		}
+		return null;
+	}
+
+	private static FlinkCompactionFilterFactory createAndSetCompactFilterFactory(@Nonnull ColumnFamilyOptions options) {
+		FlinkCompactionFilterFactory compactionFilterFactory =
+			new FlinkCompactionFilterFactory(createRocksDbNativeLogger());
+		//noinspection resource
+		options.setCompactionFilterFactory(compactionFilterFactory);
+		return compactionFilterFactory;
+	}
+
+	private static org.rocksdb.Logger createRocksDbNativeLogger() {
+		if (LOG.isDebugEnabled()) {
+			// options are always needed for org.rocksdb.Logger construction (no other constructor)
+			// the logger level gets configured from the options in native code
+			try (DBOptions opts = new DBOptions().setInfoLogLevel(InfoLogLevel.DEBUG_LEVEL)) {
+				return new org.rocksdb.Logger(opts) {
+					@Override
+					protected void log(InfoLogLevel infoLogLevel, String logMsg) {
+						LOG.debug("RocksDB filter native code log: " + logMsg);
+					}
+				};
+			}
+		} else {
+			return null;
+		}
+	}
+
+	public void configCompactFilter(
+			@Nonnull StateDescriptor<?, ?> stateDesc, FlinkCompactionFilterFactory compactionFilterFactory) {
+		StateTtlConfig ttlConfig = stateDesc.getTtlConfig();
+		boolean useSystemTime = !useCustomTimeForTtlCompactFilter &&
+			ttlConfig.getTimeCharacteristic() == StateTtlConfig.TimeCharacteristic.ProcessingTime;
+		configCompactFilter(stateDesc, compactionFilterFactory, useSystemTime);
+	}
+
+	private void configCompactFilter(
+			@Nonnull StateDescriptor<?, ?> stateDesc,
+			FlinkCompactionFilterFactory compactionFilterFactory,
+			boolean useSystemTime) {
+		StateTtlConfig ttlConfig = stateDesc.getTtlConfig();
+		if (ttlConfig.isEnabled() && ttlConfig.getCleanupStrategies().inRocksdbCompactFilter()) {
+			if (!enableTtlCompactionFilter) {
+				LOG.warn("Cannot configure RocksDB TTL compaction filter for state <{}>: " +
+					"feature is disabled for the state backend.", stateDesc.getName());
+			}
+			assert compactionFilterFactory != null;
+			FlinkCompactionFilter.Config config;
+			if (stateDesc instanceof ListStateDescriptor) {
+				ListStateDescriptor<?> listStateDesc = (ListStateDescriptor<?>) stateDesc;
+				int len = listStateDesc.getElementSerializer().getLength();
+				if (len > 0) {
+					config = FlinkCompactionFilter.Config.createForFixedElementList(
+						0, ttlConfig.getTtl().toMilliseconds(), useSystemTime, len + 1);
+				} else {
+					config = FlinkCompactionFilter.Config.createForList(
+						0, ttlConfig.getTtl().toMilliseconds(), useSystemTime,
+						new ListElementIterFactory<>(listStateDesc.getElementSerializer()));
```

Comment:
Will the `elementSerializer` be shared in c++ and java ?  Do we need to change this to `listStateDesc.getElementSerializer().duplicate()` ?